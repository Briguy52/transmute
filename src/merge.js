import curry from "./curry";
import reduce from "./reduce";
import set from "./set";

const reduceOp = reduce.operation;
const setOp = set.operation;

function merge(updates, subject) {
  return reduceOp(
    subject,
    (acc, value, key) => setOp(value, key, acc),
    updates
  );
}

/**
 * Takes each entry of `updates` and sets it on `subject`.
 *
 * @example
 * // returns Map { "one" => 3, "two" => 2, "three" => 1}
 * merge(
 *   Map({one: 1, two: 2, three: 3}),
 *   Map({one: 3, three: 1})
 * );
 *
 * @param  {Iterable} updates key-value pairs to merge in `subject`.
 * @param  {Iterable} subject the thing to update.
 * @return {Iterable} with each key-value of `updates` merged into `subject`.
 */
export default curry(merge);
