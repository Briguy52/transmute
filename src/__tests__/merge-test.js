import { Map } from "immutable";
import merge from "../merge";

describe("transmute/merge", () => {
  describe("Object", () => {
    it("merges with an Object", () => {
      expect(
        merge(
          {
            one: 3,
            three: 1
          },
          {
            one: 1,
            two: 2,
            three: 3
          }
        )
      ).toMatchSnapshot();
    });

    it("merges with a Map", () => {
      expect(
        merge(
          Map({
            one: 3,
            three: 1
          }),
          {
            one: 1,
            two: 2,
            three: 3
          }
        )
      ).toMatchSnapshot();
    });
  });

  describe("Iterable", () => {
    const merger = merge({ one: 3, three: 1 });

    it("merges with an Object", () => {
      expect(
        merger(
          Map({
            one: 1,
            two: 2,
            three: 3
          })
        )
      ).toMatchSnapshot();
    });

    it("merges with a Map", () => {
      expect(
        merger(
          Map({
            one: 1,
            two: 2,
            three: 3
          })
        )
      ).toMatchSnapshot();
    });
  });
});
