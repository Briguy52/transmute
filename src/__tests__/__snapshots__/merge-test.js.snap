// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/merge Iterable merges with a Map 1`] = `
Immutable.Map {
  one: 3,
  two: 2,
  three: 1,
}
`;

exports[`transmute/merge Iterable merges with an Object 1`] = `
Immutable.Map {
  one: 3,
  two: 2,
  three: 1,
}
`;

exports[`transmute/merge Object merges with a Map 1`] = `
Object {
  "one": 3,
  "three": 1,
  "two": 2,
}
`;

exports[`transmute/merge Object merges with an Object 1`] = `
Object {
  "one": 3,
  "three": 1,
  "two": 2,
}
`;
